### PR TITLE
Fetcher

### LINKED ISSUE #7: Add the option to download chromium bin on demand
Currenlty `chromiumoxide` relies on `chromium` being already installed on the system.
Like puppeteer an option to bundle a chromium executeable directly should be supported

### PR DESCRIPTION
**Pull request by Sytten**
Fixes #7

This is a first pass, I want to gather some feedback on it. Couple of notes:
- I used the default revision `818858` which matches the PDL `r818844` according to puppeteer (https://github.com/puppeteer/puppeteer/blob/ebd087a31661a1b701650d0be3e123cc5a813bd8/src/revisions.ts).
- I didn't implement the automatic download on launch, but I can if you guys want it.
- I had to do some magic around the feature flags because of https://github.com/rust-lang/cargo/issues/10801, I think my solution is the cleanest that we can get until it is fixed (but we do need to document it).
- I had to implement a custom extract for zip since it currently doesn't support symlinks (https://github.com/zip-rs/zip/pull/213) and those are needed on macos. I think this is cleaner than calling the OS tool (this is what `headless-chrome` does).
- I didn't add tracing into it, but I can if you guys want that.

We probably need tests for that feature, let me know what you guys would like. I think testing on all platforms should be added, but that can be in another PR.

### CODE REVIEW

File: chromiumoxide_fetcher/src/browser/runtime.rs [hunk 1]
```diff
@@ -0,0 +1,103 @@
+use std::path::{Path, PathBuf};
+
+use anyhow::Context;
+
+use super::ZipArchive;
+
+pub struct BrowserFetcherRuntime;
```
Comment (Sytten) [inline chromiumoxide_fetcher/src/browser/runtime.rs:7]: We might want an async trait for that if we want to enable both runtimes at the same time?

File: Cargo.toml [hunk 2]
```diff
@@ -43,6 +45,14 @@ tracing-subscriber = "0.2"
...
 async-std-runtime = ["async-std", "async-tungstenite/async-std-runtime"]
 tokio-runtime = ["tokio", "async-tungstenite/tokio-runtime"]
+fetcher = []
+
+# Temporary features until cargo weak dependencies bug is fixed
+# See https://github.com/rust-lang/cargo/issues/10801
+_fetcher-rustls-async-std = ["fetcher", "chromiumoxide_fetcher/_rustls-async-std"]
+_fetcher-rusttls-tokio = ["fetcher", "chromiumoxide_fetcher/_rustls-tokio"]
+_fetcher-native-async-std = ["fetcher", "chromiumoxide_fetcher/_native-async-std"]
+_fetcher-native-tokio = ["fetcher", "chromiumoxide_fetcher/_native-tokio"]
```
Comment (Sytten) [inline Cargo.toml:55]: My original plan was to have the features: `async-std-runtime`/`tokio-runtime`, `native-tls`/`rustls-tls` and `fetcher`, but I am not willing to force people to have a dirty lock file for it...

**Comment by Sytten** [comment]
@mattsse Just want to make sure the PR doesn't become stale

**Comment by mattsse** [comment]
thanks for the ping, will try to review later today.

File: chromiumoxide_fetcher/src/browser/options.rs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+impl BrowserFetcherOptions {
+    pub fn builder() -> BrowserFetcherOptionsBuilder {
+        BrowserFetcherOptionsBuilder {
+            revision: None,
+            host: None,
+            path: None,
+            platform: None,
+        }
```
Comment (mattsse) [inline chromiumoxide_fetcher/src/browser/options.rs:40]: can derive `Default`

File: chromiumoxide_fetcher/src/revision.rs [hunk 4]
```diff
@@ -0,0 +1,29 @@
+use std::fmt;
+
```
Comment (mattsse) [inline chromiumoxide_fetcher/src/revision.rs:2]:
```suggestion
```

File: chromiumoxide_fetcher/src/error.rs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+use std::num::ParseIntError;
+
```
Comment (mattsse) [inline chromiumoxide_fetcher/src/error.rs:2]:
```suggestion
```

File: chromiumoxide_fetcher/src/browser/runtime.rs [hunk 6]
```diff
@@ -0,0 +1,103 @@
+use std::path::{Path, PathBuf};
+
```
Comment (mattsse) [inline chromiumoxide_fetcher/src/browser/runtime.rs:2]:
```suggestion
```

File: chromiumoxide_fetcher/src/browser/runtime.rs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+use anyhow::Context;
+
```
Comment (mattsse) [inline chromiumoxide_fetcher/src/browser/runtime.rs:4]:
```suggestion
```

File: chromiumoxide_fetcher/src/browser/revision_info.rs [hunk 8]
```diff
@@ -0,0 +1,22 @@
+use std::fmt;
+use std::path::PathBuf;
+
```
Comment (mattsse) [inline chromiumoxide_fetcher/src/browser/revision_info.rs:3]:
```suggestion
```

File: chromiumoxide_fetcher/src/browser/options.rs [hunk 9]
```diff
@@ -0,0 +1,103 @@
+use std::path::PathBuf;
+
```
Comment (mattsse) [inline chromiumoxide_fetcher/src/browser/options.rs:2]:
```suggestion
```

File: chromiumoxide_fetcher/src/browser/options.rs [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+use directories::BaseDirs;
+
```
Comment (mattsse) [inline chromiumoxide_fetcher/src/browser/options.rs:4]:
```suggestion
```

File: chromiumoxide_fetcher/src/browser/fetcher.rs [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    pub fn new(options: BrowserFetcherOptions) -> Self {
+        Self {
+            revision: options.revision,
+            host: options.host,
+            path: options.path,
+            platform: options.platform,
+        }
+    }
+
+    pub async fn fetch(&self) -> Result<BrowserFetcherRevisionInfo> {
```
Comment (mattsse) [inline chromiumoxide_fetcher/src/browser/fetcher.rs:24]: would love some docs here

**Review by mattsse — CHANGES_REQUESTED** [review]
tysm!

this looks great.

couple of smol nits. would appreciate brief docs for the public download functions.

**Comment by Sytten** [comment]
@mattsse any reason for removing the spacing between the std, deps and crate imports? I tried to follow the rest of the codebase methodology on that.

**Comment by mattsse** [comment]
> @mattsse any reason for removing the spacing between the std, deps and crate imports? I tried to follow the rest of the codebase methodology on that.

right, I wasn't very consistent when I wrote this, the main reason is, `rusfmt` won't sort them if the imports are separated by a line

**Review by mattsse — APPROVED** [review]

**Pull request merged**